File: sdk-rust-v2/examples/02_account_dashboard.rs
//! # Account Dashboard
//!
//! Fetches account overview, positions, and open orders from the Decibel
//! testnet REST API, then displays a formatted dashboard with margin usage,
//! liquidation buffer, unrealized P&L, and order status.
//!
//! **This is a demo against the Decibel testnet. Do not use with real funds.**
//!
//! ## Environment Variables
//!
//! - `BEARER_TOKEN`         — API bearer token for the testnet.
//! - `SUBACCOUNT_ADDRESS`   — On-chain subaccount address to inspect.
//!
//! ## Usage
//!
//! ```sh
//! export BEARER_TOKEN="your_token_here"
//! export SUBACCOUNT_ADDRESS="0xYourSubaccountAddress"
//! cargo run --example 02_account_dashboard
//! ```

use decibel_sdk_v2::models::account::{AccountOverview, UserOpenOrder, UserPosition};
use decibel_sdk_v2::models::market::MarketPrice;

const BASE_URL: &str = "https://api.testnet.aptoslabs.com/decibel/api/v1";

struct Env {
    token: String,
    subaccount: String,
}

fn load_env() -> Env {
    let token = std::env::var("BEARER_TOKEN").unwrap_or_else(|_| {
        eprintln!("WARNING: BEARER_TOKEN not set — requests may fail");
        String::new()
    });
    let subaccount = std::env::var("SUBACCOUNT_ADDRESS").unwrap_or_else(|_| {
        eprintln!("WARNING: SUBACCOUNT_ADDRESS not set — using placeholder");
        "0x0000000000000000000000000000000000000000000000000000000000000000".into()
    });
    Env { token, subaccount }
}

fn client(token: &str) -> reqwest::Client {
    let mut headers = reqwest::header::HeaderMap::new();
    if !token.is_empty() {
        let val = format!("Bearer {token}");
        headers.insert(
            reqwest::header::AUTHORIZATION,
            reqwest::header::HeaderValue::from_str(&val).expect("invalid token"),
        );
    }
    reqwest::Client::builder()
        .default_headers(headers)
        .build()
        .expect("failed to build HTTP client")
}

async fn fetch_overview(
    client: &reqwest::Client,
    subaccount: &str,
) -> Option<AccountOverview> {
    let url = format!("{BASE_URL}/account/overview?subaccount={subaccount}");
    match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r.json().await.ok(),
        Ok(r) => {
            eprintln!("[overview] HTTP {}", r.status());
            None
        }
        Err(e) => {
            eprintln!("[overview] {e}");
            None
        }
    }
}

async fn fetch_positions(
    client: &reqwest::Client,
    subaccount: &str,
) -> Vec<UserPosition> {
    let url = format!("{BASE_URL}/account/positions?subaccount={subaccount}");
    match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r.json().await.unwrap_or_default(),
        Ok(r) => {
            eprintln!("[positions] HTTP {}", r.status());
            Vec::new()
        }
        Err(e) => {
            eprintln!("[positions] {e}");
            Vec::new()
        }
    }
}

async fn fetch_open_orders(
    client: &reqwest::Client,
    subaccount: &str,
) -> Vec<UserOpenOrder> {
    let url = format!("{BASE_URL}/account/orders?subaccount={subaccount}");
    match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r.json().await.unwrap_or_default(),
        Ok(r) => {
            eprintln!("[orders] HTTP {}", r.status());
            Vec::new()
        }
        Err(e) => {
            eprintln!("[orders] {e}");
            Vec::new()
        }
    }
}

async fn fetch_prices(client: &reqwest::Client) -> Vec<MarketPrice> {
    let url = format!("{BASE_URL}/prices");
    match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r.json().await.unwrap_or_default(),
        _ => Vec::new(),
    }
}

fn print_separator() {
    println!("{}", "─".repeat(72));
}

fn print_overview(overview: &AccountOverview) {
    println!("\n╔══════════════════════════════════════════════════════════════════════╗");
    println!("║                  DECIBEL TESTNET — ACCOUNT DASHBOARD                ║");
    println!("╚══════════════════════════════════════════════════════════════════════╝\n");

    println!("  {}", overview);
    println!();
    println!("  Equity:              ${:>14.2}", overview.perp_equity_balance);
    println!("  Unrealized PnL:      ${:>14.2}", overview.unrealized_pnl);
    println!("  Unrealized Funding:  ${:>14.2}", overview.unrealized_funding_cost);
    println!("  Total Margin:        ${:>14.2}", overview.total_margin);
    print_separator();
    println!(
        "  Margin Usage:        {:>14.2}%",
        overview.margin_usage_pct()
    );
    println!(
        "  Liquidation Buffer:  ${:>14.2} ({:.1}%)",
        overview.liquidation_buffer_usd(),
        overview.liquidation_buffer_pct()
    );
    println!(
        "  Withdrawable:        ${:>14.2}",
        overview.total_withdrawable()
    );

    if overview.is_liquidation_warning(25.0) {
        println!("\n  ⚠ WARNING: Liquidation buffer is below 25%!");
    }
    println!();
}

fn print_positions(positions: &[UserPosition], prices: &[MarketPrice]) {
    println!("  POSITIONS ({} active)", positions.len());
    print_separator();

    if positions.is_empty() {
        println!("  (no open positions)\n");
        return;
    }

    println!(
        "  {:<12} {:>6} {:>10} {:>12} {:>12} {:>12} {:>6}",
        "MARKET", "SIDE", "SIZE", "ENTRY", "MARK", "UNREAL PnL", "PROT"
    );

    for pos in positions {
        let mark_px = prices
            .iter()
            .find(|p| p.market == pos.market)
            .map(|p| p.mark_px)
            .unwrap_or(pos.entry_price);

        let pnl = pos.unrealized_pnl(mark_px);
        let pnl_sign = if pnl >= 0.0 { "+" } else { "" };
        let protection = if pos.has_protection() {
            "TP+SL"
        } else if pos.has_tp() {
            "TP"
        } else if pos.has_sl() {
            "SL"
        } else {
            "NONE"
        };

        println!(
            "  {:<12} {:>6} {:>10.4} {:>12.2} {:>12.2} {:>11}{:.2} {:>6}",
            pos.market,
            pos.direction().to_uppercase(),
            pos.size.abs(),
            pos.entry_price,
            mark_px,
            pnl_sign,
            pnl,
            protection,
        );
    }

    println!();

    for pos in positions {
        let mark_px = prices
            .iter()
            .find(|p| p.market == pos.market)
            .map(|p| p.mark_px)
            .unwrap_or(pos.entry_price);

        let notional = pos.notional(mark_px);
        let liq_dist = pos.liquidation_distance_pct(mark_px);

        println!("    {} | notional=${:.2} | leverage={}x | liq_price={:.2} ({:.1}% away)",
            pos.market,
            notional,
            pos.user_leverage,
            pos.estimated_liquidation_price,
            liq_dist,
        );
    }

    println!();
}

fn print_orders(orders: &[UserOpenOrder]) {
    println!("  OPEN ORDERS ({} active)", orders.len());
    print_separator();

    if orders.is_empty() {
        println!("  (no open orders)\n");
        return;
    }

    println!(
        "  {:<12} {:>6} {:>10} {:>12} {:>12} {:>10} {:>8}",
        "MARKET", "SIDE", "REMAINING", "PRICE", "NOTIONAL", "FILLED%", "TIF"
    );

    for order in orders {
        println!(
            "  {:<12} {:>6} {:>10.4} {:>12.2} {:>12.2} {:>9.1}% {:>8}",
            order.market,
            order.side().to_uppercase(),
            order.remaining_size,
            order.price,
            order.notional(),
            order.fill_pct(),
            order.time_in_force,
        );
    }

    println!();
}

#[tokio::main]
async fn main() {
    let env = load_env();
    let http = client(&env.token);

    println!("  Fetching data for subaccount: {}", env.subaccount);

    let (overview, positions, orders, prices) = tokio::join!(
        fetch_overview(&http, &env.subaccount),
        fetch_positions(&http, &env.subaccount),
        fetch_open_orders(&http, &env.subaccount),
        fetch_prices(&http),
    );

    match overview {
        Some(ref ov) => print_overview(ov),
        None => {
            println!("\n  Could not fetch account overview (check token and subaccount).\n");
        }
    }

    print_positions(&positions, &prices);
    print_orders(&orders);

    print_separator();
    println!("  Dashboard complete.\n");
}

File: decibel-sdk-python/examples/02_account_dashboard.py
#!/usr/bin/env python3
"""
Account Dashboard — monitor positions, orders, margin, and risk on Decibel testnet.

Read-only bot. No private key required.

What it does:
  1. Fetches account overview (equity, margin, PnL)
  2. Lists all open positions with unrealized PnL
  3. Lists all open orders
  4. Computes risk metrics: margin usage, liquidation distance, funding costs

Usage:
  export BEARER_TOKEN="your_bearer_token"
  export SUBACCOUNT_ADDRESS="0xyour_subaccount"
  python 02_account_dashboard.py
"""

import asyncio
import os
import sys
from datetime import datetime, timezone

import httpx

BEARER_TOKEN = os.environ.get("BEARER_TOKEN", "")
SUBACCOUNT = os.environ.get("SUBACCOUNT_ADDRESS", "")
BASE_URL = "https://api.testnet.aptoslabs.com/decibel/api/v1"
HEADERS = {
    "Authorization": f"Bearer {BEARER_TOKEN}",
    "Origin": "https://app.decibel.trade",
}


async def fetch_account_overview(client: httpx.AsyncClient) -> dict | None:
    try:
        resp = await client.get(
            f"{BASE_URL}/account_overviews",
            params={"account": SUBACCOUNT, "include_performance": "true"},
        )
        resp.raise_for_status()
        return resp.json()
    except httpx.HTTPStatusError as e:
        print(f"  Could not fetch overview: {e.response.status_code}")
        return None


async def fetch_positions(client: httpx.AsyncClient) -> list[dict]:
    try:
        resp = await client.get(
            f"{BASE_URL}/account_positions",
            params={"account": SUBACCOUNT},
        )
        resp.raise_for_status()
        return resp.json()
    except httpx.HTTPStatusError:
        return []


async def fetch_open_orders(client: httpx.AsyncClient) -> list[dict]:
    try:
        resp = await client.get(
            f"{BASE_URL}/open_orders",
            params={"account": SUBACCOUNT},
        )
        resp.raise_for_status()
        return resp.json()
    except httpx.HTTPStatusError:
        return []


async def fetch_prices(client: httpx.AsyncClient) -> dict[str, dict]:
    try:
        resp = await client.get(f"{BASE_URL}/prices")
        resp.raise_for_status()
        prices = resp.json()
        return {p.get("market", ""): p for p in prices}
    except httpx.HTTPStatusError:
        return {}


def print_overview(overview: dict | None):
    print("\n" + "=" * 70)
    print(f"  ACCOUNT DASHBOARD — {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M:%S UTC')}")
    print(f"  Subaccount: {SUBACCOUNT[:10]}...{SUBACCOUNT[-6:]}")
    print("=" * 70)

    if not overview:
        print("  No account data available (subaccount may not exist yet)")
        return

    equity = overview.get("perp_equity_balance", 0)
    margin = overview.get("total_margin", 0)
    mm = overview.get("maintenance_margin", 0)
    upnl = overview.get("unrealized_pnl", 0)
    funding_cost = overview.get("unrealized_funding_cost", 0)
    withdrawable = (
        overview.get("usdc_cross_withdrawable_balance", 0)
        + overview.get("usdc_isolated_withdrawable_balance", 0)
    )

    margin_pct = (margin / equity * 100) if equity > 0 else 0
    liq_buffer = equity - mm
    liq_buffer_pct = (liq_buffer / equity * 100) if equity > 0 else 0

    print(f"\n  {'Equity:':<25} ${equity:>12,.2f}")
    print(f"  {'Unrealized PnL:':<25} ${upnl:>+12,.2f}")
    print(f"  {'Unrealized Funding:':<25} ${funding_cost:>+12,.2f}")
    print(f"  {'Margin Used:':<25} ${margin:>12,.2f}  ({margin_pct:.1f}%)")
    print(f"  {'Maintenance Margin:':<25} ${mm:>12,.2f}")
    print(f"  {'Liquidation Buffer:':<25} ${liq_buffer:>12,.2f}  ({liq_buffer_pct:.1f}%)")
    print(f"  {'Withdrawable:':<25} ${withdrawable:>12,.2f}")

    if liq_buffer_pct < 20:
        print("\n  ⚠️  WARNING: Liquidation buffer below 20% — consider reducing exposure")
    elif liq_buffer_pct < 50:
        print("\n  ⚡ Margin is getting tight — monitor closely")


def print_positions(positions: list[dict], prices: dict[str, dict]):
    print(f"\n  OPEN POSITIONS ({len(positions)})")
    print("-" * 70)

    if not positions:
        print("  No open positions")
        return

    print(f"  {'Market':<12} {'Side':<6} {'Size':>10} {'Entry':>12} {'Mark':>12} {'PnL':>12}")
    print("  " + "-" * 64)

    for pos in positions:
        market = pos.get("market", "?")
        size = pos.get("size", 0)
        entry = pos.get("entry_price", 0)
        side = "LONG" if size > 0 else "SHORT" if size < 0 else "FLAT"
        liq = pos.get("estimated_liquidation_price", 0)

        px = prices.get(market, {})
        mark = px.get("mark_px", entry)
        pnl = (mark - entry) * size

        tp = "✓" if pos.get("tp_order_id") else "✗"
        sl = "✓" if pos.get("sl_order_id") else "✗"

        market_short = market[:10] + "..." if len(market) > 12 else market
        print(
            f"  {market_short:<12} {side:<6} {abs(size):>10.4f} "
            f"${entry:>11,.2f} ${mark:>11,.2f} ${pnl:>+11,.2f}"
        )
        print(f"  {'':>12} Liq: ${liq:,.2f}  TP:{tp} SL:{sl}")


def print_orders(orders: list[dict]):
    print(f"\n  OPEN ORDERS ({len(orders)})")
    print("-" * 70)

    if not orders:
        print("  No open orders")
        return

    print(f"  {'Market':<12} {'Side':<6} {'Price':>12} {'Size':>10} {'Filled':>8} {'Type':<8}")
    print("  " + "-" * 56)

    for order in orders:
        market = order.get("market", "?")[:12]
        side = "BUY" if order.get("is_buy") else "SELL"
        price = order.get("price", 0)
        orig = order.get("orig_size", 0)
        remaining = order.get("remaining_size", 0)
        filled_pct = ((orig - remaining) / orig * 100) if orig > 0 else 0
        tif = order.get("time_in_force", "?")

        print(
            f"  {market:<12} {side:<6} ${price:>11,.2f} {orig:>10.4f} "
            f"{filled_pct:>6.1f}% {tif:<8}"
        )


async def main():
    if not BEARER_TOKEN:
        print("Error: set BEARER_TOKEN environment variable")
        sys.exit(1)
    if not SUBACCOUNT:
        print("Error: set SUBACCOUNT_ADDRESS environment variable")
        print("Create one at https://app.decibel.trade")
        sys.exit(1)

    async with httpx.AsyncClient(headers=HEADERS, timeout=15.0) as client:
        overview, positions, orders, prices = await asyncio.gather(
            fetch_account_overview(client),
            fetch_positions(client),
            fetch_open_orders(client),
            fetch_prices(client),
        )

        print_overview(overview)
        print_positions(positions, prices)
        print_orders(orders)

    print("\n  Dashboard complete. Run again to refresh, or integrate")
    print("  into a loop with WebSocket subscriptions for live updates.")


if __name__ == "__main__":
    asyncio.run(main())
